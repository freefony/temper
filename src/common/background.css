svg {
  position: absolute;
  bottom: 0;
  top: 0;
  width: 100%;
  height: 300px;
}