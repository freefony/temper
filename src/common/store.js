import React, { Component } from 'react'

const StoreContext = React.createContext('store')

class StoreProvider extends Component {
  waiting = true
  state = {
    posts: [
      {
        title: 'No Posts available',
        body: ''
      }
    ],
    history: [],
    error: {}
  }
  
  componentDidMount() {
    this.loadPosts()
  }

  loadPosts = async() => {
    try {
      const allPosts = await (await fetch('https://jsonplaceholder.typicode.com/posts')).json()
      const posts = allPosts.slice(0, 5)
      this.setState({posts})
    } catch(error) {
      console.error(error)
      this.setState(() => ({ error }))
    }
  }

  changeHistory(title, from, to) {
    const history = [...this.state.history]
    history.unshift({title, from, to})
    this.setState(() => ({ history }))
  }

  handleTimeTravel(position) {
    const newHistory = [...this.state.history]
    newHistory.splice(0, position + 1)
    this.setState(() => ({ history: newHistory }))
  }

  changePostPosition (direction, currentIndex) {
    const { posts } = this.state
    const nextIndex = (direction === 'up') ? currentIndex - 1 : currentIndex + 1
    const rearrangedPosts = [...posts]
    const [post] = rearrangedPosts.splice(currentIndex, 1) // remove it from current position

    rearrangedPosts.splice(nextIndex, 0, post) // place it in a new position
    this.setState({posts: rearrangedPosts})
    this.changeHistory(post.title, currentIndex, nextIndex)
  }

  render() {
    const { children } = this.props
    const { posts, history } = this.state

    const store = {
      posts,
      history,
      changePostPosition: (position, currentIndex) => this.changePostPosition(position, currentIndex),
      handleTimeTravel: (position) => this.handleTimeTravel(position),
      changeHistory: (title, from, to) => this.changeHistory(title, from, to)
    }
    
    return <StoreContext.Provider value={store} >{children}</StoreContext.Provider>
  }
}

const connect = () => (Component) => (props) => {
  return <StoreContext.Consumer>
    {store => {
      const passedProps = {...props, ...store}
      return <Component {...passedProps} />
    }}
  </StoreContext.Consumer>
}

export { StoreProvider, connect }