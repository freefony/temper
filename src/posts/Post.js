import React from 'react'

import SortArrows from '../common/sort_arrow/SortArrows'

export default ({ post, onChangePosition, canMoveUp, canMoveDown, position }) => {
  return <div className="max-w-sm shadow-lg bg-white shadow rounded overflow-hidden mb-4">
    <div className="px-6 py-2 flex justify-between border">
      <p className="text-gray-700 text-base py-4">{post.title}</p>
      <SortArrows moveAction={onChangePosition} canMoveUp={canMoveUp} canMoveDown={canMoveDown} position={position} />
    </div>
  </div>
}