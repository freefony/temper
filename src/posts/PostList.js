import React  from 'react'
import Post from './Post'
import { connect } from '../common/store'
import { CSSTransition } from 'react-transition-group'

const PostList = ({ posts =[], changePostPosition }) => {
  return <div className='container'>
    <div className='text-xl text-white text-left mb-4'>Sortable Post List</div>
    {posts.map((post, i) => {
      return <CSSTransition
      unmountOnExit
      in={posts.length ? true : false}
      timeout={{ appear: 0, enter: 0, exit: 300 }}
      classNames='slide-in-down'
      appear
      key={i}
    >
      <Post 
        post={post}  
        onChangePosition={changePostPosition}
        canMoveUp={(i > 0)}
        canMoveDown={(i < posts.length - 1)}
        position={i}
      />
    </CSSTransition>
    })}
  </div>

}

export default connect()(PostList)