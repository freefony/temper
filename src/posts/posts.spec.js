import React from 'react'
import PostList from './PostList'
import Post from './Post'
import { connect } from '../common/store'
import { render, cleanup } from '@testing-library/react'

import data from '../../fixtures/posts'

describe('posts components', () => {

  afterEach(cleanup)

  describe('PostList component', () => {
    it('should render component by itself', () => {
      const { container, getByText} = render(<PostList />)
      expect(container).toMatchSnapshot()
      expect(getByText(/Sortable Post List/i)).toBeInTheDocument()
    })

    it('should render when wrapped in connect', () => {
      const PostListComp = connect()(PostList)
      const { container, getByText} = render(<PostListComp />)
      expect(container).toMatchSnapshot()
      expect(getByText(/Sortable Post List/i)).toBeInTheDocument()
    })

    it('should receive data from connect', () => {
      const _MockConnect = (Component) => (props) => {
        const responseData = data.posts.slice(0, 4)
        return <Component {...props} posts={responseData} />
      }

      const WithMockConnect = _MockConnect(PostList)
      const { container } = render(<WithMockConnect />)
      expect(container).toMatchSnapshot()
      expect(container.children[0].childNodes.length).toEqual(5)
    })
  })

  describe('Post', () => {
    it('should render post', () => {
      const { container, getByText } = render(<Post post={data.posts[0]} />)
      expect(container).toMatchSnapshot()
      expect(getByText(/arrow_up/i)).toBeInTheDocument()
      expect(getByText(/arrow_down/i)).toBeInTheDocument()
      expect(getByText(/Post 1/i)).toBeInTheDocument()
    })
    it('should not show arrow down for last element', () => {
      const { container, queryByText } = render(<Post post={data.posts[0]} canMoveDown={false} />)
      expect(container).toMatchSnapshot()
      expect(queryByText(/arrow_up/i)).toBeInTheDocument()
      expect(queryByText(/arrow_down/)).not.toBeInTheDocument()
    })
    it('should not show arrow up for first element', () => {
      const { container, queryByText } = render(<Post post={data.posts[0]} canMoveUp={false} />)
      expect(container).toMatchSnapshot()
      expect(queryByText(/arrow_down/i)).toBeInTheDocument()
      expect(queryByText(/arrow_up/)).not.toBeInTheDocument()
    })
  })
})