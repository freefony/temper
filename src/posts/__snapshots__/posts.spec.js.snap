// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`posts components Post should not show arrow down for last element 1`] = `
<div>
  <div
    class="max-w-sm shadow-lg bg-white shadow rounded overflow-hidden mb-4"
  >
    <div
      class="px-6 py-2 flex justify-between border"
    >
      <p
        class="text-gray-700 text-base py-4"
      >
        Post 1
      </p>
      <div
        class="flex flex-col justify-center"
      >
        <div
          class="arrow-btn cursor-pointer"
        >
          <svg>
            arrow_up.svg
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`posts components Post should not show arrow up for first element 1`] = `
<div>
  <div
    class="max-w-sm shadow-lg bg-white shadow rounded overflow-hidden mb-4"
  >
    <div
      class="px-6 py-2 flex justify-between border"
    >
      <p
        class="text-gray-700 text-base py-4"
      >
        Post 1
      </p>
      <div
        class="flex flex-col justify-center"
      >
        <div
          class="arrow-btn cursor-pointer"
        >
          <svg>
            arrow_down.svg
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`posts components Post should render post 1`] = `
<div>
  <div
    class="max-w-sm shadow-lg bg-white shadow rounded overflow-hidden mb-4"
  >
    <div
      class="px-6 py-2 flex justify-between border"
    >
      <p
        class="text-gray-700 text-base py-4"
      >
        Post 1
      </p>
      <div
        class="flex flex-col justify-between"
      >
        <div
          class="arrow-btn cursor-pointer"
        >
          <svg>
            arrow_up.svg
          </svg>
        </div>
        <div
          class="arrow-btn cursor-pointer"
        >
          <svg>
            arrow_down.svg
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`posts components PostList component should receive data from connect 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="text-xl text-white text-left mb-4"
    >
      Sortable Post List
    </div>
    <div
      class="max-w-sm shadow-lg bg-white shadow rounded overflow-hidden mb-4 slide-in-down-appear slide-in-down-appear-active"
    >
      <div
        class="px-6 py-2 flex justify-between border"
      >
        <p
          class="text-gray-700 text-base py-4"
        >
          Post 1
        </p>
        <div
          class="flex flex-col justify-center"
        >
          <div
            class="arrow-btn cursor-pointer"
          >
            <svg>
              arrow_down.svg
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="max-w-sm shadow-lg bg-white shadow rounded overflow-hidden mb-4 slide-in-down-appear slide-in-down-appear-active"
    >
      <div
        class="px-6 py-2 flex justify-between border"
      >
        <p
          class="text-gray-700 text-base py-4"
        >
          Post 2
        </p>
        <div
          class="flex flex-col justify-between"
        >
          <div
            class="arrow-btn cursor-pointer"
          >
            <svg>
              arrow_up.svg
            </svg>
          </div>
          <div
            class="arrow-btn cursor-pointer"
          >
            <svg>
              arrow_down.svg
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="max-w-sm shadow-lg bg-white shadow rounded overflow-hidden mb-4 slide-in-down-appear slide-in-down-appear-active"
    >
      <div
        class="px-6 py-2 flex justify-between border"
      >
        <p
          class="text-gray-700 text-base py-4"
        >
          Post 3
        </p>
        <div
          class="flex flex-col justify-between"
        >
          <div
            class="arrow-btn cursor-pointer"
          >
            <svg>
              arrow_up.svg
            </svg>
          </div>
          <div
            class="arrow-btn cursor-pointer"
          >
            <svg>
              arrow_down.svg
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="max-w-sm shadow-lg bg-white shadow rounded overflow-hidden mb-4 slide-in-down-appear slide-in-down-appear-active"
    >
      <div
        class="px-6 py-2 flex justify-between border"
      >
        <p
          class="text-gray-700 text-base py-4"
        >
          Post 4
        </p>
        <div
          class="flex flex-col justify-center"
        >
          <div
            class="arrow-btn cursor-pointer"
          >
            <svg>
              arrow_up.svg
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`posts components PostList component should render component by itself 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="text-xl text-white text-left mb-4"
    >
      Sortable Post List
    </div>
  </div>
</div>
`;

exports[`posts components PostList component should render when wrapped in connect 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="text-xl text-white text-left mb-4"
    >
      Sortable Post List
    </div>
  </div>
</div>
`;
